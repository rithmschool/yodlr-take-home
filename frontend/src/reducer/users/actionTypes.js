export const LOAD_USERS_TO_STORE = 'LOAD_USERS_TO_STORE';
export const ADD_USER_TO_STORE = 'ADD_USER_TO_STORE';
export const REMOVE_USER_FROM_STORE = 'REMOVE_USER_FROM_STORE';
export const UPDATE_USER_IN_STORE = 'UPDATE_USER_IN_STORE';