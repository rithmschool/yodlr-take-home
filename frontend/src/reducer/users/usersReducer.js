import {
  LOAD_USERS_TO_STORE, ADD_USER_TO_STORE, REMOVE_USER_FROM_STORE, UPDATE_USER_IN_STORE
} from './actionTypes';


function usersReducer(state = {}, action) {
  switch (action.type) {
    case LOAD_USERS_TO_STORE:
      const users = action.users;
      return users.reduce((table, user) => {
        table[user.id] = user;
        return table;
      }, {});

    case ADD_USER_TO_STORE:
      return { ...state, [action.user.id]: action.user };

    case REMOVE_USER_FROM_STORE:
      const data = { ...state };
      delete data[action.id];
      return data;

    case UPDATE_USER_IN_STORE:
      return { ...state, [action.user.id]: { ...action.user } };

    default:
      return state;
  }
}


export default usersReducer;