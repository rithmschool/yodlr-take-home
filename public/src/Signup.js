import React, { useState } from 'react';
import { Form, FormGroup, Button, Input } from 'reactstrap';

function Signup ({ signup }) {
    const initialState = {firstName: "", lastName: "", email: ""}
    const [formData, setFormData] = useState(initialState)

    // handle user form input before submit
    function handleChange (evt) {
        const { name, value } = evt.target;
        setFormData(fData => ({
          ...fData,
          [name]: value
        }));
    }
    
    // handle form submission
    async function handleSubmit (evt) {
        evt.preventDefault();
        try{
        await signup(formData);
        setFormData(initialState);
        } catch (err) {
            return( <p>Bad login, please try again.</p>)
        }
    }


    // display signup form
    return (
        <div>
            <h2>Sign up for updates about our company!</h2>
            <Form onSubmit={handleSubmit}>
                <FormGroup>
                    <Input type="text" name="firstName" placeholder="First name" onChange={handleChange} required></Input>
                </FormGroup>
                <FormGroup>
                    <Input type="text" name="lastName" placeholder="Last name" onChange={handleChange} required></Input>
                </FormGroup>
                <FormGroup>
                    <Input type="text" name="email" placeholder="Email" onChange={handleChange} required></Input>
                </FormGroup>
                <Button>Sign up</Button>
            </Form>
        </div>
    )
}

export default Signup;