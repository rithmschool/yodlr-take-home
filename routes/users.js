var express = require('express');
var router = express.Router();
var _ = require('lodash');
var logger = require('../lib/logger');
var log = logger();
const { ensureAdmin } = require('../helpers/auth.js');

var users = require('../init_data.json').data;
var curId = _.size(users);

/* GET users listing. */
router.get('/', ensureAdmin, function(req, res) {
  return res.json(_.toArray(users));
});

/* Create a new user */
router.post('/', function(req, res) {
  var user = req.body;
  user.id = curId++;
  if (!user.state) {
    user.state = 'pending';
  }
  users[user.id] = user;
  log.info('Created user', user);
  return res.json(user);
});

/* Get a specific user by id */
router.get('/:id', ensureAdmin, function(req, res, next) {
  var user = users[req.params.id];
  if (!user) {
    return next();
  }
  return res.json(users[req.params.id]);
});

/* Delete a user by id */
router.delete('/:id', ensureAdmin, function(req, res) {
  var user = users[req.params.id];
  delete users[req.params.id];
  res.status(204);
  log.info('Deleted user', user);
  return res.json(user);
});

/* Update a user by id */
router.put('/:id', ensureAdmin, function(req, res, next) {
  var user = req.body;
  if (user.id != req.params.id.id) {
    return next(new Error('ID paramter does not match body'));
  }
  users[user.id] = user;
  log.info('Updating user', user);
  return res.json(user);
});


module.exports = router;
