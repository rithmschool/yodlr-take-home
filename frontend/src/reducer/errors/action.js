import { SET_ERROR, RESET_ERROR } from './actionTypes';


function setError(message) {
  return { type: SET_ERROR, message };
}

function resetError() {
  return { type: RESET_ERROR };
}


export { setError, resetError };