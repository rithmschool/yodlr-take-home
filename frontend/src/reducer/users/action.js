import {
  LOAD_USERS_TO_STORE, ADD_USER_TO_STORE, REMOVE_USER_FROM_STORE, UPDATE_USER_IN_STORE
} from './actionTypes';
import User from '../../api/User';
import { setError, resetError } from '../errors/action';


function storeUser(user) {
  return { type: ADD_USER_TO_STORE, user };
}

/**
 * Send post request to backend and add new user to redux store.
 * @param {Object} user - { email, firstName, lastName, state }
 */
function addUser(user) {
  return async function(dispatch) {
    dispatch(resetError());

    try {
      const newUser = await User.create(user);
      dispatch(storeUser(newUser));
    } catch (error) {
      console.error(error);
      setError(error.message);
    }
  }
}

function removeUserFromStore(id) {
  return { type: REMOVE_USER_FROM_STORE, id };
}

/**
 * Sends DELETE request to backend and remove user from redux store.
 * @param {Number} id 
 */
function removeUser(id) {
  return async function(dispatch) {
    dispatch(resetError());

    try {
      await User.remove(id);
      dispatch(removeUserFromStore(id));
    } catch (error) {
      console.error(error);
      dispatch(setError(error.message));
    }
  }
}

function loadUsersToStore(users) {
  return { type: LOAD_USERS_TO_STORE, users };
}

/**
 * Sends GET request to backend and loads users to redux store.
 */
function loadUsers() {
  return async function(dispatch) {
    dispatch(resetError());

    try {
      const users = await User.getAll();
      dispatch(loadUsersToStore(users));
    } catch (error) {
      console.error(error);
      dispatch(setError(error.message));
    }
  }
}

function updateStoredUser(user) {
  return { type: UPDATE_USER_IN_STORE, user };
}

function updateUser(user) {
  return async function(dispatch) {
    dispatch(resetError());

    try {
      await User.update(user.id, user)
      dispatch(updateStoredUser(user));
    } catch (error) {
      console.error(error);
      dispatch(setError(error.message));
    }
  }
}


export { addUser, removeUser, updateUser, loadUsers };