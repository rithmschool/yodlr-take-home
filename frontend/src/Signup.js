import React, { useState } from 'react';
import { useDispatch, useSelector } from 'react-redux';
import { addUser } from './reducer/users/action'
import { useHistory } from 'react-router-dom';


const Signup = () => {
  const error = useSelector(state => state.error);
  const INIT_FORM_STATE = {
    "firstName": "",
    "lastName": "",
    "email": "",
  }
  const [formData, setFormData] = useState(INIT_FORM_STATE);
  const [formError, setFormError] = useState("");
  const dispatch = useDispatch();
  const history = useHistory();

  const handleChange = (evt) => {
    const { name, value } = evt.target;
    setFormData(f => ({ ...f, [name]: value }));
  }

  const handleSubmit = (evt) => {
    evt.preventDefault();
    try {
      dispatch(addUser({ ...formData, state: "pending" }));
      setFormData(INIT_FORM_STATE);
      history.push('/');
    } catch (e) {
      console.error(e);
      setFormError(error);
    }
  }


  return (
    <div className="row justify-content-center mt-5">
      <div className="card col-md-6">
        <form className="card-body text-left" onSubmit={handleSubmit}>
          <div className="form-row">
            <div className="form-group col-md-6">
              <label htmlFor="intput-first-name">First Name</label>
              <input type="text" className="form-control" id="intput-first-name"
                name="firstName" value={formData.firstName} onChange={handleChange} />
            </div>
            <div className="form-group col-md-6">
              <label htmlFor="intput-last-name">Last Name</label>
              <input type="text" className="form-control" id="intput-last-name"
                name="lastName" value={formData.lastName} onChange={handleChange} />
            </div>
          </div>
          <div className="form-group">
            <label htmlFor="input-email">Email</label>
            <input type="text" className="form-control" id="input-email"
              name="email" value={formData.email} onChange={handleChange} />
          </div>
          <button type="submit" className="btn btn-primary btn-block"
            disabled={Object.values(formData).some(field => field.length === 0)}>
            Sign up
          </button>
          {formError.length !== 0 &&
            <div className="alert alert-danger" role="alert">
              {formError}
            </div>
          }
        </form>
      </div>
    </div>
  )
}


export default Signup;