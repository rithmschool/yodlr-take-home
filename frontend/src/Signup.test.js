import React from 'react';
import { MemoryRouter } from 'react-router-dom';
import { useSelector, useDispatch } from 'react-redux';
import { getByLabelText, getByText, render, fireEvent } from '@testing-library/react';
import { addUser } from './reducer/users/action'
import Signup from './Signup';


jest.mock('react-redux', () => ({
  useSelector: jest.fn(),
  useDispatch: jest.fn()
}));

jest.mock('./reducer/users/action', () => ({
  addUser: jest.fn(),
}))

beforeEach(function() {
  useSelector.mockImplementation((selectorFn) =>
    selectorFn({
      users: {
        0: {
          id: 0, firstName: "test", lastName: "user", state: "active"
        },
        1: {
          id: 1, firstName: "test1", lastName: "user", state: "pending"
        }
      }
    })
  );
});

// smoke test
it("renders without crashing", function() {
  render(<MemoryRouter><Signup /></MemoryRouter>);
});

// snapshot test
it("matches snapshot", function() {
  const { asFragment } = render(<MemoryRouter><Signup /></MemoryRouter>);
  expect(asFragment()).toMatchSnapshot();
});

describe("form values can be update", function() {
  it("first name input", function() {
    const { getByLabelText } = render(<MemoryRouter><Signup /></MemoryRouter>);
    const firstNameInput = getByLabelText("First Name");

    fireEvent.change(firstNameInput, { target: { value: 'test first name' } });
    expect(firstNameInput.value).toBe("test first name");
  });

  it("last name input", function() {
    const { getByLabelText } = render(<MemoryRouter><Signup /></MemoryRouter>);
    const lastNameInput = getByLabelText("Last Name");

    fireEvent.change(lastNameInput, { target: { value: 'test last name' } });
    expect(lastNameInput.value).toBe("test last name");
  });

  it("email input", function() {
    const { getByLabelText } = render(<MemoryRouter><Signup /></MemoryRouter>);
    const emailInput = getByLabelText("Email");

    fireEvent.change(emailInput, { target: { value: 'test email' } });
    expect(emailInput.value).toBe("test email");
  });
});

it("Dispatches addUser() on submit", function() {
  const mockedDispatch = jest.fn();
  useDispatch.mockReturnValue(mockedDispatch);
  const { getByLabelText, getByText } = render(<MemoryRouter><Signup /></MemoryRouter>);

  const firstNameInput = getByLabelText("First Name");
  const lastNameInput = getByLabelText("Last Name");
  const emailInput = getByLabelText("Email");
  fireEvent.change(firstNameInput, { target: { value: 'test first name' } });
  fireEvent.change(lastNameInput, { target: { value: 'test last name' } });
  fireEvent.change(emailInput, { target: { value: 'test email' } });

  const submitBtn = getByText("Sign up");
  fireEvent.click(submitBtn);

  expect(mockedDispatch).toHaveBeenCalledWith(addUser({
    firstName: "test first name",
    lastName: "test last name",
    email: "test email"
  }));
});