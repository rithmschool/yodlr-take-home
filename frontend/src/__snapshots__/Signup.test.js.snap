// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="row justify-content-center mt-5"
  >
    <div
      class="card col-md-6"
    >
      <form
        class="card-body text-left"
      >
        <div
          class="form-row"
        >
          <div
            class="form-group col-md-6"
          >
            <label
              for="intput-first-name"
            >
              First Name
            </label>
            <input
              class="form-control"
              id="intput-first-name"
              name="firstName"
              type="text"
              value=""
            />
          </div>
          <div
            class="form-group col-md-6"
          >
            <label
              for="intput-last-name"
            >
              Last Name
            </label>
            <input
              class="form-control"
              id="intput-last-name"
              name="lastName"
              type="text"
              value=""
            />
          </div>
        </div>
        <div
          class="form-group"
        >
          <label
            for="input-email"
          >
            Email
          </label>
          <input
            class="form-control"
            id="input-email"
            name="email"
            type="text"
            value=""
          />
        </div>
        <button
          class="btn btn-primary btn-block"
          disabled=""
          type="submit"
        >
          Sign up
        </button>
      </form>
    </div>
  </div>
</DocumentFragment>
`;
